engines/ospray/ispc/geometry/ExtendedCones.ispc

```diff
@@ -32,70 +32,68 @@
+    float centerRadius;
+    float upRadius;
+    float timestamp;
+    vec2f texture_coords;
+    int32 materialID;
+};
+
+__define_safe_index(ExtendedCone);
+
 struct ExtendedCones
 {
     uniform Geometry geometry;
 
-    uniform uint8 *uniform data;
+    uniform ExtendedCone* uniform data;
 
-    float radius;
-    float length;
     int materialID;
-    int offset_center;
-    int offset_up;
-    int offset_centerRadius;
-    int offset_upRadius;
-    int offset_timestamp;
-    int offset_texture_coords;
-    int offset_materialID;
     int32 numExtendedCones;
-    int32 bytesPerCone;
+    uniform bool useSafeIndex;
 };
 
-void ExtendedCones_bounds(uniform ExtendedCones *uniform geometry,
-                          uniform size_t primID, uniform box3fa &bbox)
+void ExtendedCones_bounds(uniform ExtendedCones* uniform geometry,
+                          uniform size_t primID, uniform box3fa& bbox)
 {
-    uniform uint8 *uniform conePtr =
-        geometry->data + geometry->bytesPerCone * primID;
-    uniform float extent = geometry->radius;
-    if (geometry->offset_centerRadius >= 0)
-        extent = *((uniform float *)(conePtr + geometry->offset_centerRadius));
+    uniform ExtendedCone* uniform conePtr =
+        safe_index(geometry->useSafeIndex, geometry->data, primID);
 
-    if (geometry->offset_upRadius >= 0)
-    {
-        uniform float upRadius =
-            *((uniform float *)(conePtr + geometry->offset_upRadius));
-        if (upRadius > extent)
-            extent = upRadius;
-    }
-    uniform vec3f v0 = *((uniform vec3f *)(conePtr + geometry->offset_center));
-    uniform vec3f v1 = *((uniform vec3f *)(conePtr + geometry->offset_up));
+    uniform float extent = conePtr->centerRadius;
+    uniform float upRadius = conePtr->upRadius;
+
+    if (upRadius > extent)
+        extent = upRadius;
+
+    uniform vec3f v0 = conePtr->center;
+    uniform vec3f v1 = conePtr->up;
     bbox = make_box3fa(min(v0, v1) - make_vec3f(extent),
                        max(v0, v1) + make_vec3f(extent));
 }
 
-void ExtendedCones_intersect(uniform ExtendedCones *uniform geometry,
-                             varying Ray &ray, uniform size_t primID)
+void ExtendedCones_intersect(uniform ExtendedCones* uniform geometry,
+                             varying Ray& ray, uniform size_t primID)
 {
-    uniform uint8 *uniform conePtr =
-        geometry->data + geometry->bytesPerCone * primID;
-
-    uniform float radius0 = geometry->radius;
-    if (geometry->offset_centerRadius >= 0)
-        radius0 = *((uniform float *)(conePtr + geometry->offset_centerRadius));
+    uniform ExtendedCone* uniform conePtr =
+        safe_index(geometry->useSafeIndex, geometry->data, primID);
 
-    uniform float radius1 = geometry->radius;
-    if (geometry->offset_upRadius >= 0)
-        radius1 = *((uniform float *)(conePtr + geometry->offset_upRadius));
+    uniform float radius0 = conePtr->centerRadius;
+    uniform float radius1 = conePtr->upRadius;
```

Comment:
I think the performance is very similar since the only difference is the radius is read from the cone instead of the global geometry. Also, the whole cone geometry is always uploaded anyways so the same memory is used.